artifactory/commands/terraform/terraformpublish.go

```diff
@@ -0,0 +1,304 @@
+			}
+			isTerraformModule, e := checkIfTerraformModule(path)
+			if e != nil {
+				return e
+			}
+			if isTerraformModule {
+				uploadParams, e := ts.uploadParamsForTerraformPublish(pathInfo.Name(), strings.TrimPrefix(path, pwd+string(filepath.Separator)))
+				if e != nil {
+					return e
+				}
+				dataHandlerFunc := services.GetSaveTaskInContentWriterFunc(toArchive, *uploadParams, errorsQueue)
+				e = services.CollectFilesForUpload(*uploadParams, nil, nil, dataHandlerFunc)
+				if e != nil {
+					return e
+				}
+				// SkipDir will not stop the walk, but will jump to the next directory.
+				return filepath.SkipDir
+			}
+			return nil
+		})
+		if err != nil && err != io.EOF {
+			log.Error(err)
+			errorsQueue.AddError(err)
+		}
+
+		// Upload modules
+		for targetPath, archiveData := range toArchive {
+			err := archiveData.GetWriter().Close()
+			if err != nil {
+				log.Error(err)
+				errorsQueue.AddError(err)
+			}
+			// Upload module using upload service
+			serviceManager, err := utils.CreateServiceManager(ts.serverDetails, 0, 0, false)
+			if err != nil {
+				log.Error(err)
+				errorsQueue.AddError(err)
+			}
+			uploadService := services.NewUploadService(serviceManager.Client())
+			uploadService.ArtDetails = serviceManager.GetConfig().GetServiceDetails()
+			uploadService.Threads = serviceManager.GetConfig().GetThreads()
+			producer.AddTaskWithError(uploadService.CreateUploadAsZipFunc(uploadSummary, targetPath, archiveData, errorsQueue), errorsQueue.AddError)
+		}
+	}()
+}
+
+func (ts *TerraformPublishCommand) performTerraformPublishTasks(consumer parallel.Runner, uploadSummary *clientservicesutils.Result) (totalUploaded, totalFailed int) {
+	// Blocking until consuming is finished.
+	consumer.Run()
+	totalUploaded = clientservicesutils.SumIntArray(uploadSummary.SuccessCount)
+	totalUploadAttempted := clientservicesutils.SumIntArray(uploadSummary.TotalCount)
+
+	log.Debug("Uploaded", strconv.Itoa(totalUploaded), "artifacts.")
+	totalFailed = totalUploadAttempted - totalUploaded
+	if totalFailed > 0 {
+		log.Error("Failed uploading", strconv.Itoa(totalFailed), "artifacts.")
+	}
+	return
+}
+
+func (tp *TerraformPublishCommand) uploadParamsForTerraformPublish(moduleName, dirPath string) (*services.UploadParams, error) {
+	uploadParams := services.NewUploadParams()
+	uploadParams.Target = tp.getPublishTarget(moduleName)
+	uploadParams.Pattern = dirPath + "/(*)"
+	uploadParams.TargetPathInArchive = "{1}"
+	uploadParams.Archive = "zip"
+	uploadParams.Recursive = true
+	uploadParams.CommonParams.TargetProps = specutils.NewProperties()
+	uploadParams.CommonParams.Exclusions = append(tp.exclusions, "*.git", "*.DS_Store")
+
+	return &uploadParams, nil
+}
+
+// Module's path in terraform repository : namespace/provider/moduleName/tag.zip
+func (tp *TerraformPublishCommand) getPublishTarget(moduleName string) string {
+	return filepath.ToSlash(filepath.Join(tp.repo, tp.namespace, tp.provider, moduleName, tp.tag+".zip"))
```

Comment:
I think that instead of using ```filepath.Join``` and then correcting its slashes with ```filepath.ToSlash```, you could have just used ```path.Join```.
The difference between ```filepath.Join``` and ```path.Join``` is that the first one, will use backslashes foe Windows, and slashes for unix, and the second one will always use slashes, which is exactly what you need.